src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.ReadCore.cs

```diff
@@ -0,0 +1,113 @@
+                        }
+                        else
+                        {
+                            // Read more data until we have the full element.
+                            state.BytesConsumed += reader.BytesConsumed;
+                            return default!;
+                        }
+                    }
+                }
+                else
+                {
+                    // For a continuation, read ahead here to avoid having to build and then tear
+                    // down the call stack if there is more than one buffer fetch necessary.
+                    if (!SingleValueReadWithReadAhead(ClassType.Value, ref reader, ref state))
+                    {
+                        state.BytesConsumed += reader.BytesConsumed;
+                        return default!;
+                    }
+                }
+
+                JsonPropertyInfo jsonPropertyInfo = state.Current.JsonClassInfo.PolicyProperty!;
+                bool success = TryRead(ref reader, jsonPropertyInfo.RuntimePropertyType!, options, ref state, out T value);
+                if (success)
+                {
+                    // Read any trailing whitespace. This will throw if JsonCommentHandling=Disallow.
+                    // Avoiding setting ReturnValue for the final block; reader.Read() returns 'false' even when this is the final block.
+                    if (!reader.Read() && !reader.IsFinalBlock)
+                    {
+                        // This method will re-enter if so set `ReturnValue` which will be returned during re-entry.
+                        state.Current.ReturnValue = value;
+                    }
+                }
+
+                state.BytesConsumed += reader.BytesConsumed;
+                return value;
+            }
+            catch (JsonReaderException ex)
+            {
+                ThrowHelper.ReThrowWithPath(state, ex);
+                return default;
+            }
+            catch (FormatException ex) when (ex.Source == ThrowHelper.ExceptionSourceValueToRethrowAsJsonException)
+            {
+                ThrowHelper.ReThrowWithPath(state, reader, ex);
+                return default;
+            }
+            catch (InvalidOperationException ex) when (ex.Source == ThrowHelper.ExceptionSourceValueToRethrowAsJsonException)
+            {
+                ThrowHelper.ReThrowWithPath(state, reader, ex);
+                return default;
+            }
+            catch (JsonException ex)
+            {
+                ThrowHelper.AddJsonExceptionInformation(state, reader, ex);
+                throw;
+            }
+            catch (NotSupportedException ex)
+            {
+                // $.RealJsonProperty.ClrPropertyName
+                // Type "int[,] is not supported. Parent CLR Type: MyPoco. Path: rootprop.prop2 (real json path)
+
+                NotSupportedException newEx = ThrowHelper.GetNotSupportedException(state, reader, ex);
+                if (newEx == ex)
```

Comment:
It can't be a helper method since "throw" without argument needs to be within the `catch`.

Reply:
But you can re-throw `ex` argument inside `GetNotSupportedException` if `message.Contains(" Path: ")`; otherwise throw the new exception.

If you do that, you can just leave the call to ThrowHelper.GetNotSupportedException(state, reader, ex) in this catch block.

See suggestion: https://github.com/dotnet/runtime/pull/32669/files#r388716071

Reply:
Per other comment "throw ex" will not have same semantics as "throw" since "throw ex" creates a new call stack which is not desired.

Reply:
It makes sense now, disregard this suggestion.